### ISSUE TITLE
Use named type parameters for traits

### ISSUE THREAD

**Issue by clarus** [formal-land/coq-of-rust#77]
With traits, it is sometimes unclear what type parameter is referring to. Instead of using explicit ordered type parameters, we should use implicit named types parameters with curly braces.

**Comment by clarus** [comment]
We can take the file https://github.com/formal-land/coq-of-rust/blob/main/examples/generics/generics_traits.rs as an example to translate with named parameters, as it contains a trait with type parameters:
```rust
// Non-copyable types.
struct Empty;
struct Null;

// A trait generic over `T`.
trait DoubleDrop<T> {
    // Define a method on the caller type which takes an
    // additional single parameter `T` and does nothing with it.
    fn double_drop(self, _: T);
}

// Implement `DoubleDrop<T>` for any generic parameter `T` and
// caller `U`.
impl<T, U> DoubleDrop<T> for U {
    // This method takes ownership of both passed arguments,
    // deallocating both.
    fn double_drop(self, _: T) {}
}

fn main() {
    let empty = Empty;
    let null = Null;

    // Deallocate `empty` and `null`.
    empty.double_drop(null);

    //empty;
    //null;
    // ^ TODO: Try uncommenting these lines.
}
```

Here is a translation that we can make:
```coq
(* Generated by coq-of-rust *)
Require Import CoqOfRust.CoqOfRust.

Import Root.std.prelude.rust_2015.

Module Empty.
  Inductive t : Set := Build.
End Empty.
Definition Empty := Empty.t.

Module Null.
  Inductive t : Set := Build.
End Null.
Definition Null := Null.t.

Module DoubleDrop.
  Class Trait {Self T : Set} : Set := {
    double_drop : Self -> T -> (M unit);
  }.
  
  Global Instance Method_double_drop `(Trait) : Notation.Dot "double_drop" := {
    Notation.dot := double_drop;
  }.
End DoubleDrop.

Module Impl_DoubleDrop_for_U.
Section Impl_DoubleDrop_for_U.
  Context {U T : Set}.

  Definition Self := U.
  
  Definition double_drop (self : Self) (Pattern : T) : M unit := Pure tt.
  
  Global Instance Method_double_drop : Notation.Dot "double_drop" := {
    Notation.dot := double_drop;
  }.
  
  Global Instance I : DoubleDrop.Trait (Self := U) (T := T) := {
    DoubleDrop.double_drop := double_drop;
  }.
End Impl_DoubleDrop_for_U.
End Impl_DoubleDrop_for_U.

(* #[allow(dead_code)] - function was ignored by the compiler *)
Definition main (_ : unit) : M unit :=
  let empty := Empty.Build in
  let null := Null.Build in
  let* _ := empty.["double_drop"] null in
  Pure tt.
```

The current translation (that does not type-check) is the following:
```coq
(* Generated by coq-of-rust *)
Require Import CoqOfRust.CoqOfRust.

Import Root.std.prelude.rust_2015.

Module Empty.
  Inductive t : Set := Build.
End Empty.
Definition Empty := Empty.t.

Module Null.
  Inductive t : Set := Build.
End Null.
Definition Null := Null.t.

Module DoubleDrop.
  Class Trait (T Self : Set) : Set := {
    double_drop : Self -> T -> (M unit);
  }.
  
  Global Instance Method_double_drop `(Trait) : Notation.Dot "double_drop" := {
    Notation.dot := double_drop;
  }.
End DoubleDrop.

Module Impl_DoubleDrop_for_U.
  Definition Self := U.
  
  Definition double_drop (self : Self) (Pattern : T) : M unit := Pure tt.
  
  Global Instance Method_double_drop : Notation.Dot "double_drop" := {
    Notation.dot := double_drop;
  }.
  
  Global Instance I T U : DoubleDrop.Trait Self T := {
    DoubleDrop.double_drop := double_drop;
  }.
End Impl_DoubleDrop_for_U.

(* #[allow(dead_code)] - function was ignored by the compiler *)
Definition main (_ : unit) : M unit :=
  let empty := Empty.Build in
  let null := Null.Build in
  let* _ := empty.["double_drop"] null in
  Pure tt.
```

What this new translation would bring is:
* curly braces around the trait parameters for the definition: `Class Trait {Self T : Set} : Set := {`
* a section to add polymorphic type variables for the trait implementation:
  ```coq
  Section Impl_DoubleDrop_for_U.
    Context {U T : Set}.
  ```
* naming the trait parameters when instantiating the trait: `Global Instance I : DoubleDrop.Trait (Self := U) (T := T) := {`

**Comment by InfiniteEchoes** [comment]
> curly braces around the trait parameters for the definition: Class Trait {Self T : Set} : Set := {
For all parameters?

**Comment by clarus** [comment]
> For all parameters?

OK we can do it also for all the parameters except `Self`, if that is what you had in mind.

**Comment by clarus** [comment]
Closed by #100
